internal/api/mfa.go

```diff
@@ -61,6 +69,74 @@ const (
 	QRCodeGenerationErrorMessage   = "Error generating QR Code"
 )
 
+func (a *API) enrollSMSFactor(w http.ResponseWriter, r *http.Request, user *models.User, session *models.Session, params *EnrollFactorParams) error {
+	ctx := r.Context()
+	config := a.config
+	db := a.db.WithContext(ctx)
+	if !config.MFA.SMS.EnrollEnabled {
+		return unprocessableEntityError(ErrorCodeMFASMSEnrollDisabled, "mfa enroll is disabled for sms")
+	}
+	if params.PhoneNumber == "" {
+		return badRequestError(ErrorCodeValidationFailed, "Phone number required to enroll SMS factor")
+	}
+
+	phone, err := validatePhone(params.PhoneNumber)
+	if err != nil {
+		return badRequestError(ErrorCodeValidationFailed, "Invalid phone number format (E.164 required)")
+	}
+	factors := user.Factors
+
+	factorCount := len(factors)
+	numVerifiedFactors := 0
+	if err := models.DeleteExpiredFactors(db, config.MFA.FactorExpiryDuration); err != nil {
+		return err
+	}
+
+	for _, factor := range factors {
+		if factor.IsVerified() {
+			numVerifiedFactors += 1
+		}
+	}
+
+	if factorCount >= int(config.MFA.MaxEnrolledFactors) {
+		return forbiddenError(ErrorCodeTooManyEnrolledMFAFactors, "Maximum number of verified factors reached, unenroll to continue")
+	}
+
+	if numVerifiedFactors >= config.MFA.MaxVerifiedFactors {
+		return forbiddenError(ErrorCodeTooManyEnrolledMFAFactors, "Maximum number of verified factors reached, unenroll to continue")
+	}
+
+	if numVerifiedFactors > 0 && !session.IsAAL2() {
+		return forbiddenError(ErrorCodeInsufficientAAL, "AAL2 required to enroll a new factor")
+	}
+	factor := models.NewSMSFactor(user, phone, params.FriendlyName, params.FactorType, models.FactorStateUnverified)
+	err = db.Transaction(func(tx *storage.Connection) error {
+		if terr := tx.Create(factor); terr != nil {
+			pgErr := utilities.NewPostgresError(terr)
+			if pgErr.IsUniqueConstraintViolated() {
+				return unprocessableEntityError(ErrorCodeMFAFactorNameConflict, fmt.Sprintf("A factor with the friendly name %q for this user likely already exists", factor.FriendlyName))
+			}
+			return terr
+
+		}
```

Comment:
Why do we rely on the constraint here? Why not select and see if there's a factor with the name already?

Reply:
Also possible - we can also iterate over existing factors w/o doing a database check. This is a copy from existing check from MFA (TOTP)

Reply:
OK let's refactor this in a new PR. Not a fan of constraints as those can be easily dropped by users (though this is not encouraged).